crates/iceberg/src/writer/base_writer/data_file_writer.rs

```diff
@@ -0,0 +1,323 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module provide `DataFileWriter`.
+
+use crate::spec::{DataContentType, DataFile, Struct};
+use crate::writer::file_writer::FileWriter;
+use crate::writer::CurrentFileStatus;
+use crate::writer::{file_writer::FileWriterBuilder, IcebergWriter, IcebergWriterBuilder};
+use crate::Result;
+use arrow_array::RecordBatch;
+use itertools::Itertools;
+
+/// Builder for `DataFileWriter`.
+#[derive(Clone)]
+pub struct DataFileWriterBuilder<B: FileWriterBuilder> {
+    inner: B,
+}
+
+impl<B: FileWriterBuilder> DataFileWriterBuilder<B> {
+    /// Create a new `DataFileWriterBuilder` using a `FileWriterBuilder`.
+    pub fn new(inner: B) -> Self {
+        Self { inner }
+    }
+}
+
+/// Config for `DataFileWriter`.
+pub struct DataFileWriterConfig {
+    partition_value: Struct,
+}
+
+impl DataFileWriterConfig {
+    /// Create a new `DataFileWriterConfig` with partition value.
+    pub fn new(partition_value: Option<Struct>) -> Self {
+        Self {
+            partition_value: partition_value.unwrap_or(Struct::empty()),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<B: FileWriterBuilder> IcebergWriterBuilder for DataFileWriterBuilder<B> {
+    type R = DataFileWriter<B>;
+    type C = DataFileWriterConfig;
+
+    async fn build(self, config: Self::C) -> Result<Self::R> {
+        Ok(DataFileWriter {
+            inner_writer: self.inner.clone().build().await?,
+            builder: self.inner,
+            partition_value: config.partition_value,
+        })
+    }
+}
+
+/// A writer write data is within one spec/partition.
+pub struct DataFileWriter<B: FileWriterBuilder> {
+    builder: B,
+    inner_writer: B::R,
+    partition_value: Struct,
+}
+
+#[async_trait::async_trait]
+impl<B: FileWriterBuilder> IcebergWriter for DataFileWriter<B> {
+    async fn write(&mut self, batch: RecordBatch) -> Result<()> {
+        self.inner_writer.write(&batch).await
+    }
+
+    async fn flush(&mut self) -> Result<Vec<DataFile>> {
+        let writer = std::mem::replace(&mut self.inner_writer, self.builder.clone().build().await?);
```

Comment:
Hi, I'm bit confused about this. Why we need this? What will happen if user calling `flush` on the same file twice?

Reply:
Because the FileWriter only provides the interface `close(self)` which will make it simple. To provide the interface `flush(&mut self)`, each time flush we will create a new FileWriter. 

A FileWriter will write the data into one or multiple files. There is no restriction for that. 

>  What will happen if user calling flush on the same file twice?

It's safe. The semantic of flush is to flush data written before into the storage and generate the files.

Reply:
I'm thinking about the error handling cases. If user hits error during flush, they can't retry it and will lost all existing writing data. How about pushing the writer back if it flush failed? User can decide whether abort the whole writing process or retry this flush.

Reply:
> I'm thinking about the error handling cases. If user hits error during flush, they can't retry it and will lost all existing written data. How about pushing the writer back if it flush failed? User can decide whether abort the whole writing process or retry this flush.

It's fine if you think it's out of the current scope. We can create an issue for this and keep moving forward.

Reply:
> I'm thinking about the error handling cases. If user hits error during flush, they can't retry it and will lost all existing written data. How about pushing the writer back if it flush failed? User can decide whether abort the whole writing process or retry this flush.

Thanks! I think you remind me that this way is not good for the error handle. 
For a flush process, there are two phases:
1. close inner writer.
2. create a new writer.
Because of `fn close(self)`, we have to replace a new writer before closing successfully. So there are two error cases that may happen:
1. create a new writer fail so that we skip to close the original writer 
2. close fail but we create a new writer
I think there are two solutions:
1. Use `fn close(&mut self)` 
2. Use `Option<Writer>` so that we can take the writer temporarily.

Reply:
I does prefer to use `fn close(&mut self)` since `close` also might return error.

Reply:
> I'm thinking about the error handling cases. If user hits error during flush, they can't retry it and will lost all existing written data. How about pushing the writer back if it flush failed? User can decide whether abort the whole writing process or retry this flush.

I find that we can't guarantee that the user can retry the flush. E.g. [the close interface for parquet ](https://docs.rs/parquet/latest/parquet/arrow/async_writer/struct.AsyncArrowWriter.html#method.close) will consume itself. I think a more simple semantic we can guarantee that if flush fails, the data will lose all existing written data and the writer can be used again. It is not friendly for users but it's easier to maintain its semantics.
For this semantic, I think `Option<Writer> and fn close(self)` may be more appropriate because it guarantees that a writer will not be used again if the close fails.

Reply:
> I think a more simple semantic we can guarantee that if flush fails, the data will lose all existing written data and the writer can be used again. It is not friendly for users but it's easier to maintain its semantics.

Makes sense. By the way, opendal can retry internally, so this should not be a big issue. Let's keep moving.